Bug 1383252: Reduce the number of hg clone operations

mathjazz opened the pull request:
This is 3 things and it's better to look at them commit by commit:

1. Purely cosmetic. But it helps a little with the following ones.
2. If `sync_projects` is ran with the `--locale` parameter, we only sync the specified locale, but we also pull repositories for all other locales within the project. And rely on that fact in a few places. This anomaly is now gone, which and helps with the final commit.
3. The actual bugfix. More info in commit description.

mathjazz commented on pontoon/sync/core.py:
```diff
@@ -241,3 +252,74 @@ def commit_changes(db_project, vcs_project, changeset, locale):
+                    response.raise_for_status()
+
+                # If locale has changed in the VCS, we need to sync it.
+                latest_commit_id = get_json_value_by_path(response.json(), repo.latest_commit_path)
```
I wonder if there's a better way of solving this problem.

Currently we have two repository sources for Firefox (and a few other projects):
* https://api.bitbucket.org/2.0/repositories/mozilla-l10n/arn/commits
* https://hg.mozilla.org/l10n-central/ja/rev/default?style=json

Since we need to retrieve commit hash from 2 different JSON files, I added this `latest_commit_path` which tries to simulate what XPath does.

Pike replied:
The API endpoints should be:

https://hg.mozilla.org/l10n-central/de/json-rev/default -> get result.node
https://api.bitbucket.org/2.0/repositories/mozilla-l10n/arn/commit/default -> get result.hash
https://api.github.com/repos/mozilla-l10n/firefoxios-l10n/git/refs/heads/master -> get result.object.sha

Note the rate limit for gh, we might need to get an API key, or add that in a separate bug.

mathjazz replied:
I believe we should limit this optimization to the 6 projects that use per-locale repositories, which means we don't need to change anything for the GitHub repositories. In > 90% of cases all 27 of those are done in less than a minute.

With that, I wonder if I should drop `get_json_value_by_path` and change `latest_commit_path` to `latest_commit_key`?

Pike replied:
As with the API end points, I think this shouldn't be in the DB, and depending on how we abstract the hosting API, this might just move into that abstraction.

flodolo commented on pontoon/sync/core.py:
```diff
@@ -241,3 +252,74 @@ def commit_changes(db_project, vcs_project, changeset, locale):
+
+                # If locale has changed in the VCS, we need to sync it.
+                latest_commit_id = get_json_value_by_path(response.json(), repo.latest_commit_path)
+                if latest_commit_id[:12] != repo.get_last_synced_revisions(locale.code):
```
Curiosity: isn't it standard to use the first characters of a commit hash?

mathjazz replied:
Yes. This should be the same as `latest_commit_id[0:12]`. I could change for clarity.

Pike replied:
.startswith is OK, but we should aim to use full-length shas for this.

flodolo replied:
Not sure why I read this the other way around…

mathjazz commented:
Added another commit that I'll squash with the 3rd one after the review.

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -98,40 +97,35 @@ def sync_project(
+    # Sync source strings. We cannot sync sources if locale is specified,
+    # because that would apply any source string changes to the specified locale only.
+    if locale:
+        source_changes = None
```
Make it an empty dict here? ...

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -98,40 +97,35 @@ def sync_project(
-        resource_changes['new_paths'],
+        source_changes['project_changes'] if source_changes else None,
+        source_changes['obsolete_vcs_resources'] if source_changes else None,
+        source_changes['new_paths'] if source_changes else None,
```
... then you can just .get() here.

mathjazz replied:
Mmmm, I love this!

Pike commented on pontoon/base/models.py:
```diff
@@ -1321,6 +1321,16 @@ class Repository(models.Model):
+    latest_commit_path = models.CharField(max_length=2000, blank=True, help_text="""
+        A slash-separated list of keys and indices used to retrieve the latest commit hash value
+        from the `api_endpoint` JSON response.
+    """)
```
This one bothers me from a technical POV.

We shouldn't trust admins to know and get API endpoints right. I think we should do something programmatically to transform the repo to API.

flodolo replied:
+1 on this. Setting up a new project would be a bunch of "open another project, copy and paste setting from there".

mathjazz replied:
PMs should not bother setting those up. That's why I'm hiding these fields from the currently used and documented project admin interface. I can also hide them from the Django admin.

But I'm not exactly sure if we can automate this.

Pike replied:
This shouldn't be hard to do programmatically, it's just a regular expression replacement for each hosting service.

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -246,6 +234,20 @@ def sync_translations(
 
+    # If project repositories have API access, we can retrieve latest commit hashes and detect
+    # changed locales before the expensive VCS pull/clone operations. When performing full scan and
+    # when source strings have changed, we still need to sync all locales.
```
Or do we? I think we do for a full scan, but for en-US changes, I think we don't. We used to when we sync'ed removed strings and templates, but as we're only syncing files with l10n activity today, we don't?

mathjazz replied:
Yeah, when I woke up I remembered I needed to add this check and while I was driving to gym I realized I remembered wrong. Will fix.

Pike commented on pontoon/base/models.py:
```diff
@@ -1401,6 +1391,37 @@ def can_commit(self):
+        url = self.url
+
+        if url.startswith('ssh://hg.mozilla.org/'):
+            endpoint_base = url.replace('ssh://', 'https://')
```
I find this a bit hard to digest. How about

```
'https://{0.netloc}{0.path}json-rev/default'.format(urlparse(r.url))
```

jotes replied:
@Pike :+1: Optionally, we could do something like this:
```
url = urlparse(self.url)
# That dict can be moved into constant somewhere above.
supported_repos =  {
  'hg.mozilla.org': {
       'endpoint': 'https://hg.mozilla.org/json-rev/default',
      'node': 'node',
   },
  (...)
}
if url.netloc in supported_repos:
   repo = supported_repos[url.netloc]
   repo['endpoint'] = repo['endpoint'].format(url)
   return repo
else:
   return None
```
That could reduce number of nested code blocks in this property in the future and even we could make this more reusable for non-mozilla related projects (?) <insert philosophical raptor here>

Pike commented on pontoon/base/models.py:
```diff
@@ -1401,6 +1391,37 @@ def can_commit(self):
+            endpoint_base = url.replace('ssh://', 'https://')
+            return {
+                'endpoint': os.path.join(endpoint_base, 'json-rev/default'),
+                'key': 'node',
```
Can we return a lambda function instead?

That way we'll be able to easily add gh API support in the future.

jotes replied:
Alternative option: https://pypi.python.org/pypi/jsonpath/ . However, that may be overkill.

Pike commented on pontoon/base/models.py:
```diff
@@ -1401,6 +1391,37 @@ def can_commit(self):
+                'key': 'node',
+            }
+
+        elif url.startswith('ssh://hg@bitbucket.org/'):
```
omg, what a nitpicker :-(

No else after return. Also below.

Pike commented on pontoon/base/models.py:
```diff
@@ -1401,6 +1391,37 @@ def can_commit(self):
+            endpoint_base = url.replace(
+                'ssh://hg@bitbucket.org/',
+                'https://api.bitbucket.org/2.0/repositories/',
+            )
```
```
https://api.bitbucket.org/2.0/repositories/{0.path}/commit/default'.format(urlparse(url))
```
would work here.

Pike commented on pontoon/sync/core.py:
```diff
@@ -264,7 +263,7 @@ def get_changed_locales(db_project, locales):
     # Requirement: all translation repositories must have API configured.
     for repo in repos:
-        if not (repo.api_endpoint and repo.latest_commit_path):
+        if not repo.api_config:
```
I'd cache the api_config here, and not use the getter twice below.

jotes commented on pontoon/base/models.py:
```diff
@@ -1401,6 +1391,37 @@ def can_commit(self):
+                'https://api.bitbucket.org/2.0/repositories/',
+            )
+            return {
+                'endpoint': os.path.join(endpoint_base, 'commit/default'),
```
Do we need to do os.path.join here? I would move `commit/default` to the rest of url at 1415.

jotes commented on pontoon/sync/core.py:
```diff
@@ -241,3 +251,74 @@ def commit_changes(db_project, vcs_project, changeset, locale):
+                response = requests.get(locale_api_endpoint)
+
+                # Raise exception on 4XX client error or 5XX server error response
+                if response.status_code != 200:
```
that isn't necessary for response.raise_for_status()?

mathjazz replied:
Nice catch!

mathjazz commented:
I restored production DB on stage and synced Firefox (96 locales) with and without this patch applied. In both cases there were 5 locales that needed to be synced.

Wihtout patch:
* total sync time: 16:10
* hg time (clone + identify): 15:26

With patch:
* total sync time: 01:54
* hg time (clone and id): 00:46
* api time: 00:34

Pike commented on pontoon/sync/core.py:
```diff
@@ -241,3 +251,85 @@ def commit_changes(db_project, vcs_project, changeset, locale):
+            # Errors and exceptions can mean locale is in a different repository or indicate
+            # an actual network problem.
+            except requests.exceptions.RequestException as e:
+                log.info(
```
This sounds a bit spammy over time? Like, we'll get this each sync process for all locales on bitbucket from hgmo?

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -229,11 +224,18 @@ def sync_translations(
+    # changed locales before the expensive VCS pull/clone operations. When performing full scan,
+    # we still need to sync all locales.
+    if not full_scan:
+        locales = get_changed_locales(db_project, locales) or locales
```
This will sync all locales if it should sync none.

This goes back to the changed_locales code returning None if any of the repos doesn't have an API. Maybe we should revert that logic and sync all locales for which we don't have an API return code, similar to what we're doing for http errors?

Then this `or locales` clause would work via get_changed_locales directly, and the "no locales need syncing" bug would be fixed.

Pike commented:
> when source strings change, we need to pull all locales and sync them to update stats

Maybe we need to "sync" them, but we don't need to pull them?

mathjazz commented:
> Maybe we need to "sync" them, but we don't need to pull them?

That's the problem - we currently need files for sync, so we need to pull.

Should be a relatively simple fix.

mathjazz commented:
I reverted the temporary commit and added an actual fix to properly sync locales on project changes, even if we don't pull them.

It's a little bigger than what I was hoping for. I made a few changes unnecessary for this fix, but they should make the code more readable.

@Pike Since you're the reviewer of this PR, I wonder if you could review this commit, too?

mathjazz commented on pontoon/sync/core.py:
```diff
@@ -165,6 +165,26 @@ def update_translated_resources(db_project, vcs_project, locale):
                 translatedresource.calculate_stats()
 
 
+def update_translated_resources_no_files(db_project, locale, changed_resources):
```
This is the main change. It updates translated resources for all sources that have source string changed.

mathjazz commented on pontoon/sync/core.py:
```diff
@@ -165,6 +165,27 @@ def update_translated_resources(db_project, vcs_project, locale):
+            log.error(
+                'Unable to calculate stats for asymmetric resource: {resource}.'.format(resource)
+            )
+            continue
```
This should never happen, because bilingual files need to be updated for all locales (merged), when the source change.

Still, maybe we should throw an exception here instead, so that we notice the issue?

Pike replied:
What would happen on next sync if we do?

mathjazz replied:
Regardless of raising an exception, we no longer have the information about the project changes in the following sync. That's [bug 1383271](https://bugzilla.mozilla.org/show_bug.cgi?id=1383271).

Pike replied:
I'd not error out here for now. Also, I'm not 100% confident that we should restrict ourselves to have projects with only monolingual or only bilingual files.

mathjazz replied:
We have projects with a mix of monolingual or bilingual files (e.g. Fundraising). They should work fine with this code.

If the source .lang file changes, all locales will be updated in the repository, so we'll have files for all locales and use the current `update_translated_resources` method.

If only the source .properties file changes, we'll use `update_translated_resources_no_files` for locales that don't have any locale changes.

mathjazz commented on pontoon/sync/tasks.py:
```diff
@@ -238,14 +240,17 @@ def sync_translations(
+        changed_resources = db_project.resources.filter(
+            Q(entities__date_created=now) |
+            Q(entities__pk__in=updated_entity_pks + obsolete_entity_pks)
         )
```
* `project_changes['update_db']`: entities to update in DB, if e.g. string or comment changes and key doesn't
* `project_changes['obsolete_db']`: entities to obsolete
* `project_changes['create_db']`: entities to create in DB

Since `project_changes['create_db']` doesn't contain PKs, we need to fetch those entities using `Q(entities__date_created=now)`. All newly created entities have the same timestamp as the sync (`now`).

Alternative approach would be to gather `changed_resources` while syncing sources and pass them all the way here, but I'd rather do this as a followup to this PR (to avoid making it even bigger).

Pike replied:
The 'now' query has a race condition with regular translations, and other sync process, right?

Do we use bulk_insert? https://docs.djangoproject.com/en/1.11/ref/models/querysets/#bulk-create says that starting with 1.10, on postgres, we get the IDs back.

mathjazz replied:
> The 'now' query has a race condition with regular translations, and other sync process, right?

No. Regular translations are stored in a different model (Translation) than source strings (Entity). Other sync processes for the same project [cannot happen at the same time](https://github.com/mozilla/pontoon/blob/master/pontoon/sync/core.py#L37).

> Do we use bulk_insert? https://docs.djangoproject.com/en/1.11/ref/models/querysets/#bulk-create says that starting with 1.10, on postgres, we get the IDs back.

No, but we [should](https://github.com/mozilla/pontoon/blob/master/pontoon/sync/changeset.py#L167). Thanks. Although that wouldn't bring us IDs here.

Pike replied:
'k

mathjazz commented on pontoon/sync/tasks.py:
```diff
@@ -283,34 +289,13 @@ def sync_translations(
                 if not vcs_project.synced_locales:
                     vcs_project.resources
 
-                if not obsolete_vcs_resources:
```
This check is actually no longer needed. It's role was to prevent skipping the commit when files were removed from source repository - so we could also remove them from locale repository. But we no longer do that.

And, we should rename `obsolete_vcs_resources` to `removed_paths` in a followup.

mathjazz commented on pontoon/sync/tasks.py:
```diff
@@ -283,34 +289,13 @@ def sync_translations(
-                        continue
+                # Skip all locales if none of the them has anything to sync
+                if len(vcs_project.synced_locales) == 0:
+                    break
```
And now that we also update stats after the big `for locale in locales` loop (to c), we can simply break instead of return.

mathjazz commented on pontoon/sync/tasks.py:
```diff
@@ -283,34 +289,13 @@ def sync_translations(
+
+                # Skip locales that have nothing to sync
+                if vcs_project.synced_locales and locale not in vcs_project.synced_locales:
+                    continue
```
Similarly as above, stats for those locales will be updated at the end of the big for loop.

Pike commented on pontoon/sync/core.py:
```diff
@@ -165,6 +165,27 @@ def update_translated_resources(db_project, vcs_project, locale):
 
+def update_translated_resources_no_files(db_project, locale, changed_resources):
+    """
+    Create/update TranslatedResource entries if files aren't available. This tipically happens when
```
'tipi' -> 'typically'

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -238,14 +240,17 @@ def sync_translations(
-            project_changes['update_db'] +
-            project_changes['obsolete_db'] +
-            project_changes['create_db']
+        updated_entity_pks = [x[1].pk for x in project_changes['update_db']]
```
Can you use a real tuple here that hints at why we're taking the second field?

Pike commented on pontoon/sync/tasks.py:
```diff
@@ -244,7 +244,10 @@ def sync_translations(
-        updated_entity_pks = [x[1].pk for x in project_changes['update_db']]
+        updated_entity_pks = []
+        for locale_code, db_entity, vcs_entity in project_changes['update_db']:
+            updated_entity_pks = db_entity.pk
```
.append(db_entity.pk)
